### PR TITLE
Python 2.7 and Python 3.6 Compatibility

### PR DESCRIPTION
**Pull request by npalmer-professional**
This commit makes the vast majority of HARK compatible with both Python 2.7 and Python 3.6. There extensive commit notes on what was done [here](https://github.com/npalmer-professional/HARK-1/commit/ebe301f602dd98f982fff9e4f9d04c8c60c5acae).

### CODE REVIEW

**Comment by npalmer-professional** [comment]
Hey all, I just wanted to quickly add some formal reviewers. I don't know if "reviewers" or "assignees" is the best thing to use for this -- also don't know how many should be requested. I figure we just need 1-2 other "good sets of eyes" on it. @mgerring You are welcome to look as well, couldn't figure out how to add you as a possible reviewer.

**Comment by llorracc** [comment]
Jason,
    For various things I am doing, it would be convenient if this request could be reviewed and approved (or sent back for improvements) in the next couple of days.  I tried to add Matthew Gerring as a potential alternative reviewer but for some reason didn't seem to be able to do that, but if the two of you could confer (also maybe with Sumana) and figure out a strategy that would be helpful.

-- Chris

File: HARK/cAndCwithStickyE/StickyEtools.py [hunk 1]
```diff
@@ -185,7 +189,7 @@ def makeStickyEdataFile(Economy,ignore_periods,description='',filename=None,save
                 DataArray = np.hstack((DataArray,np.reshape(R,(R.size,1))))
                 VarNames.append('R')
             with open(results_dir + filename + 'Data.txt','wb') as f:
-                my_writer = csv.writer(f, delimiter = '\t')
+                my_writer = csv.writer(f, delimiter = r'\t')
```
Comment (jasonaowen) [inline HARK/cAndCwithStickyE/StickyEtools.py]: In particular, I don't think this does what you want it to do - it looks like it will write a CSV file with a delimiter of the character `\` followed by the character `t`, rather than a tab character.

File: HARK/cAndCwithStickyE/StickyEtools.py [hunk 2]
```diff
@@ -532,76 +536,76 @@ def makeResultsPanel(Coeffs,StdErrs,Rsq,Pvals,OID,Counts,meas_err,sticky,all_spe
...
-    memo = ''
+    memo = r''
     if all_specs:
-        memo += '\\\\ \multicolumn{6}{l}{Memo: For instruments $\mathbf{Z}_{t}$, ' + DeltaLogC + ' $= \mathbf{Z}_{t} \zeta, ~\\bar{R}^{2}=$ ' + mystr1(Counts[3])
+        memo += r'\\\\ \multicolumn{6}{l}{Memo: For instruments $\mathbf{Z}_{t}$, ' + DeltaLogC + ' $= \mathbf{Z}_{t} \zeta, ~\\bar{R}^{2}=$ ' + mystr1(Counts[3])
         if meas_err:
-            memo += '; ~$\\var(\\log(\\xi_t))=$ ' + mystr3(Counts[4])
-        memo += ' }  \n'
+            memo += r'; ~$\\var(\\log(\\xi_t))=$ ' + mystr3(Counts[4])
+        memo += r' }  \n'
```
Comment (jasonaowen) [inline HARK/cAndCwithStickyE/StickyEtools.py]: Same raw string literal concern here: this newline character just turned into a backslash followed by the letter `n`. Is that what you wanted?

File: HARK/core.py [hunk 3]
```diff
@@ -939,7 +950,15 @@ def solveAgents(self):
...
         #    this_type.solve()
-        multiThreadCommands(self.agents,['solve()'])
+        try:
+            multiThreadCommands(self.agents,['solve()'])
+        except Exception as err:
+            if self.print_parallel_error_once:
+                # Set flag to False so this is only printed once. 
+                self.print_parallel_error_once = False
+                print("**** WARNING: could not execute multiThreadCommands in HARK.core.Market.solveAgents(), so using the serial version instead. This will likely be slower. The multiTreadCommands() functions failed with the following error:", '\n    ', sys.exc_info()[0], ':', err) #sys.exc_info()[0])
+            multiThreadCommandsFake(self.agents,['solve()'])
```
Comment (jasonaowen) [inline HARK/core.py]: This is a nice solution! Graceful degradation is good.

It might be worth putting it in its own commit.

File: HARK/cstwMPC/cstwMPC.py [hunk 4]
```diff
@@ -9,7 +16,8 @@
                            getPercentiles, getLorenzShares, calcSubpopAvg, approxLognormal
 from HARK.simulation import drawDiscrete
 from HARK import Market
-import SetupParamsCSTW as Params
+#import SetupParamsCSTW as Params
```
Comment (jasonaowen) [inline HARK/cstwMPC/cstwMPC.py]: Please don't commit commented-out code. :)

File: HARK/estimation.py [hunk 5]
```diff
@@ -5,11 +5,13 @@
...
 # The following libraries are part of the standard python distribution
 from __future__ import division                         # Use new division function
+from __future__ import print_function
+from builtins import str
 import numpy as np                                      # Numerical Python
 from time import time                                   # Used to time execution
 from copy import deepcopy                               # For replicating complex objects
 from scipy.optimize import fmin, fmin_powell            # Minimizers
-from utilities import warnings                          # Import modified "warnings" library
+#from utilities import warnings                          # Import modified "warnings" library
```
Comment (jasonaowen) [inline HARK/estimation.py]: Please don't commit commented-out code. :)

File: HARK/parallel.py [hunk 6]
```diff
@@ -3,28 +3,40 @@
...
+# such that they will raise useful errors if called.
+def raiseImportError(moduleStr):
+    def defineImportError(*args, **kwargs):
+        raise ImportError(moduleStr + ' could not be imported, and is required for this'+\
+        ' function.  See HARK documentation for more information on how to install the ' \
+        + moduleStr + ' module.')
+    return defineImportError
+
 try:
-    # Try to import joblib and dill
```
Comment (jasonaowen) [inline HARK/parallel.py]: :+1: nice fix!

File: HARK/cAndCwithStickyE/StickyEtools.py [hunk 7]
```diff
@@ -13,29 +17,29 @@
...
 from HARK.interpolation import LinearInterp
-from StickyEparams import results_dir, tables_dir, figures_dir, UpdatePrb, PermShkAggVar
+from .StickyEparams import results_dir, tables_dir, figures_dir, UpdatePrb, PermShkAggVar
 UpdatePrbBase = UpdatePrb
 PermShkAggVarBase = PermShkAggVar
 
 def mystr1(number):
     if not np.isnan(number):
-        out = "{:.3f}".format(number)
+        out = r"{:.3f}".format(number)
```
[the remark this answers is not in the document]
Comment (npalmer-professional) [reply HARK/cAndCwithStickyE/StickyEtools.py]: Yeah I think I found another solution, will remove these literals and replace with a better solution.

File: HARK/cAndCwithStickyE/StickyEtools.py (hunk 1 above)
Comment (npalmer-professional) [reply HARK/cAndCwithStickyE/StickyEtools.py]: Yes, this has been fixed.

File: HARK/ConsumptionSaving/ConsAggShockModel.py [hunk 8]
```diff
@@ -1749,8 +1752,8 @@ def __init__(self,AFunc):
 
 ###############################################################################
 
-if __name__ == '__main__':
-    import ConsumerParameters as Params
+def main():
```
[the remark this answers is not in the document]
Comment (npalmer-professional) [reply HARK/ConsumptionSaving/ConsAggShockModel.py]: I've added this and committed on my fork; will rebase and squash these commit togther.

File: HARK/SolvingMicroDSOPs/EstimationParameters.py [hunk 9]
```diff
@@ -2,6 +2,23 @@
...
+# Error Notes:
+# This sets a "quick fix" to the error, AttributeError: 'TempConsumerType' object has no attribute 'PermGroFacAgg'
+# If you set this flag to "True" you will see the error. A more thorough fix is to
+# fix the place where this error was introduced (Set to "True" and it will appear;  
+# this was almost certainly introduced when the code was extended to be used in the 
+# GE setting). An even more thorough solution, which moves beyond the scope of 
+# fixing this error, is adding unit tests to ID when changes to some code will 
+# break things elsewhere. 
+# Note: alternatively, decide that the "init_consumer_objects['PermGroFacAgg'] = 1.0"
+# line below fixes it properly ('feature not a bug') and remove all this text.
```
[the remark this answers is not in the document]
Comment (npalmer-professional) [reply HARK/SolvingMicroDSOPs/EstimationParameters.py]: In the future I will make commits like these their own small commits.

File: HARK/cstwMPC/cstwMPC.py (hunk 4 above)
Comment (npalmer-professional) [reply HARK/cstwMPC/cstwMPC.py]: Removed!

File: HARK/estimation.py (hunk 5 above)
Comment (npalmer-professional) [reply HARK/estimation.py]: Removed!

**Comment by npalmer-professional** [comment]
@jasonaowen Hey Jason, we have a wrinkle to resolve before we fianlly merge. It looks like `builtins` is not part of the "plain" anaconda 2.7 install -- `builtins` is in `future` which is not in "plain anaconda." We want this version of HARK to work "out of the box" for Py2 and Py3 under anaconda, so we're going to need to figure out the right solution here. I *think* the right solution is the following -- let me know what you think: 
1. add `future` to the requirements for pip/conda install-able HARK  (so users who pip/conda-install HARK will automatically have `future` and `builtins` automatically)
2. for users who have directly downloaded the code, if we want that to still run, I *think* the only solution is to have "try/except" clauses around the "builtins" imports, and track down the couple places that absolutely require "builtins" (such as next() notation) and add a try/except there as well.  Hmmm.  Would love to know your thoughts!

**Comment by npalmer-professional** [comment]
I've spoke with @jasonaowen and we were able to work through his questions. We decided that I would do a final run through of tests after I implemented the changes we talked about. I've completed that today and am merging this. One of our next steps will be a build for pip/conda install under Python 2.7 and 3.
